File: python-elasticemail-examples/examples/ee.py
"""
Shared configuration for the Elastic Email examples.

Every script does `from ee import ...` after adding its own directory to sys.path,
so they can be run from the project root: python examples/basic_send.py
"""

import hashlib
import hmac
import math
import os
import sys
import threading
import time
from urllib.parse import urlparse

import ElasticEmail
from dotenv import load_dotenv

load_dotenv()

FROM = os.environ.get("EMAIL_FROM", "Acme <hello@yourdomain.com>")
TO = os.environ.get("EMAIL_TO", "you@yourdomain.com")
CONTACT_EMAIL = os.environ.get("CONTACT_EMAIL", FROM)
LIST_NAME = os.environ.get("ELASTICEMAIL_LIST_NAME", "Newsletter")
TEMPLATE_NAME = os.environ.get("ELASTICEMAIL_TEMPLATE_NAME", "welcome-example")
PUBLIC_URL = os.environ.get("PUBLIC_URL", "http://localhost:3000")
WEBHOOK_TOKEN = os.environ.get("ELASTICEMAIL_WEBHOOK_TOKEN", "change_me")
SENDING_DOMAIN = os.environ.get("SENDING_DOMAIN", "yourdomain.com")
CONFIRM_REDIRECT_URL = os.environ.get("CONFIRM_REDIRECT_URL", "")
# Required by POST /send. Callers send it as "Authorization: Bearer <token>".
SEND_TOKEN = os.environ.get("ELASTICEMAIL_SEND_TOKEN", "")
CONFIRM_LINK_TTL = 48 * 3600


def get_configuration():
    """Build an SDK Configuration from ELASTICEMAIL_API_KEY or exit with a clear message."""
    api_key = os.environ.get("ELASTICEMAIL_API_KEY")
    if not api_key:
        print(
            "ELASTICEMAIL_API_KEY is not set. Copy .env.example to .env and add your key "
            "from https://app.elasticemail.com/marketing/settings/new/manage-api",
            file=sys.stderr,
        )
        sys.exit(1)

    configuration = ElasticEmail.Configuration()
    configuration.api_key["apikey"] = api_key
    return configuration


def print_api_error(step, e):
    """Print status and body of an ElasticEmail.ApiException."""
    print("Error ({}): {} {}".format(step, e.status, e.body), file=sys.stderr)


def api_error_message(e):
    """Extract the Error string from an ApiException body, falling back to the raw body."""
    import json

    try:
        return json.loads(e.body).get("Error", e.body)
    except (TypeError, ValueError):
        return e.body or str(e)


# ---------------------------------------------------------------------------
# Security helpers shared by the server apps
# ---------------------------------------------------------------------------

_HTML_ESCAPES = str.maketrans(
    {"&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;", "{": "&#123;", "}": "&#125;"}
)


def escape_html(value):
    """Escape a user-supplied value for an HTML body.

    Elastic Email treats `{...}` and `{{...}}` in message content as template syntax,
    so braces are escaped as well.
    """
    return str(value if value is not None else "").translate(_HTML_ESCAPES)


def plain_text(value):
    """Strip braces from a user-supplied value for PlainText bodies and contact fields."""
    return str(value if value is not None else "").replace("{", "").replace("}", "")


def header_text(value):
    """plain_text() plus no CR/LF, for Subject and other header-like fields."""
    return plain_text(value).replace("\r", "").replace("\n", "")


def safe_equal(a, b):
    """Constant-time string comparison (bytes, so non-ASCII input cannot raise)."""
    return hmac.compare_digest(str(a or "").encode(), str(b or "").encode())


def email_shape_ok(email):
    """Reject anything with whitespace, CR/LF, or not exactly one @."""
    if not isinstance(email, str) or email.count("@") != 1:
        return False
    return not any(c.isspace() for c in email)


def _allowed_domains():
    domains = [d.strip().lower() for d in os.environ.get("EMAIL_ALLOWED_DOMAINS", "").split(",")]
    domains = [d for d in domains if d]
    if not domains:
        email_to = os.environ.get("EMAIL_TO", "")
        if "@" in email_to:
            default = email_to.rsplit("@", 1)[1].strip().lower()
            if default:
                domains = [default]
    return set(domains)


ALLOWED_DOMAINS = _allowed_domains()


def recipient_allowed(email):
    """True when the recipient's domain is in EMAIL_ALLOWED_DOMAINS (default: the domain of EMAIL_TO)."""
    if not email_shape_ok(email):
        return False
    return email.rsplit("@", 1)[1].lower() in ALLOWED_DOMAINS


def send_token_ok(authorization):
    """Authorization must be exactly "Bearer <ELASTICEMAIL_SEND_TOKEN>"."""
    header = str(authorization or "")
    if not header.startswith("Bearer "):
        return False
    return safe_equal(header[len("Bearer "):], SEND_TOKEN)


def sign_confirm(secret, email, expires):
    """HMAC-SHA256 of email + "\n" + expires, hex encoded."""
    return hmac.new(secret.encode(), "{}\n{}".format(email, expires).encode(), hashlib.sha256).hexdigest()


def confirm_url(public_url, secret, email):
    """Double opt-in confirmation link, valid for 48 hours."""
    from urllib.parse import quote

    expires = int(time.time()) + CONFIRM_LINK_TTL
    return "{}/double-optin/confirm?email={}&expires={}&token={}".format(
        public_url, quote(email, safe=""), expires, sign_confirm(secret, email, expires)
    )


def confirm_link_ok(secret, email, expires, token):
    """Check a confirmation link: all fields present, not expired, HMAC matches."""
    if not email or not expires or not token:
        return False
    expires = str(expires)
    if not (expires.isascii() and expires.isdigit()) or int(expires) < time.time():
        return False
    return safe_equal(sign_confirm(secret, email, expires), token)


class FixedWindowLimiter:
    """Fixed-window rate limiter.

    In-memory limits are per process; use a shared store (Redis, the platform's rate limiter)
    in production.
    """

    def __init__(self, limit, window_seconds):
        self.limit = limit
        self.window = window_seconds
        self.hits = {}
        self.lock = threading.Lock()

    def hit(self, key):
        """Count one request. Returns None when allowed, or seconds until the window resets."""
        now = time.time()
        with self.lock:
            if len(self.hits) > 10000:
                self.hits = {k: v for k, v in self.hits.items() if v[0] + self.window > now}
            start, count = self.hits.get(key, (now, 0))
            if start + self.window <= now:
                start, count = now, 0
            if count >= self.limit:
                return max(1, math.ceil(start + self.window - now))
            self.hits[key] = (start, count + 1)
            return None


# POST /double-optin/subscribe: per client IP 5 per 10 minutes, per email 3 per hour
SUBSCRIBE_IP_LIMIT = FixedWindowLimiter(5, 10 * 60)
SUBSCRIBE_EMAIL_LIMIT = FixedWindowLimiter(3, 60 * 60)


def subscribe_rate_limited(ip, email):
    """Returns None when allowed, or the Retry-After seconds."""
    return SUBSCRIBE_IP_LIMIT.hit(str(ip or "")) or SUBSCRIBE_EMAIL_LIMIT.hit(email.lower())


def check_startup(webhook_token=None, public_url=None, send_route=True):
    """Startup checks for the long-running servers."""
    if webhook_token is None:
        webhook_token = WEBHOOK_TOKEN
    if public_url is None:
        public_url = os.environ.get("PUBLIC_URL", "")
    if send_route and not SEND_TOKEN:
        print("POST /send is disabled until ELASTICEMAIL_SEND_TOKEN is set", file=sys.stderr)
    if webhook_token in ("", "change_me"):
        host = urlparse(public_url).hostname if public_url else None
        if not public_url or host in ("localhost", "127.0.0.1", "::1"):
            print("ELASTICEMAIL_WEBHOOK_TOKEN is the placeholder; fine for local testing only", file=sys.stderr)
        else:
            print(
                "Refusing to start: set ELASTICEMAIL_WEBHOOK_TOKEN before exposing webhooks at {}".format(public_url),
                file=sys.stderr,
            )
            sys.exit(1)

File: go-elasticemail-examples/internal/ee/ee.go
// Package ee holds the configuration shared by every example: env loading,
// the Elastic Email client with its API key context, and error printing.
package ee

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"math"
	"net/http"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"

	ElasticEmail "github.com/elasticemail/elasticemail-go/v4"
	"github.com/joho/godotenv"
)

// LoadEnv reads .env from the current directory if present. Missing file is fine.
func LoadEnv() {
	_ = godotenv.Load()
}

// Env returns the variable value or the fallback when unset or empty.
func Env(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

// NewClient builds the API client and the context carrying the API key.
// The SDK reads the key from the context on every request and sends it as
// the X-ElasticEmail-ApiKey header.
func NewClient() (*ElasticEmail.APIClient, context.Context) {
	LoadEnv()

	apiKey := os.Getenv("ELASTICEMAIL_API_KEY")
	if apiKey == "" {
		log.Fatal("ELASTICEMAIL_API_KEY environment variable is required. Get a key at https://app.elasticemail.com/marketing/settings/new/manage-api")
	}

	ctx := context.WithValue(
		context.Background(),
		ElasticEmail.ContextAPIKeys,
		map[string]ElasticEmail.APIKey{
			"apikey": {Key: apiKey},
		},
	)

	client := ElasticEmail.NewAPIClient(ElasticEmail.NewConfiguration())
	return client, ctx
}

// From is the verified sender address.
func From() string {
	return Env("EMAIL_FROM", "Acme <hello@yourdomain.com>")
}

// To is the test recipient.
func To() string {
	return Env("EMAIL_TO", "you@yourdomain.com")
}

// ErrorBody extracts the raw response body from an SDK error. The API answers
// failures with {"Error": "message"}; the SDK keeps that body on GenericOpenAPIError.
func ErrorBody(err error) string {
	var apiErr *ElasticEmail.GenericOpenAPIError
	if errors.As(err, &apiErr) {
		body := strings.TrimSpace(string(apiErr.Body()))
		if body != "" {
			return body
		}
	}
	if err != nil {
		return err.Error()
	}
	return ""
}

// StatusCode returns the HTTP status of a response, or 0 when there is none.
func StatusCode(resp *http.Response) int {
	if resp == nil {
		return 0
	}
	return resp.StatusCode
}

// PrintAPIError writes "Error (<step>): <status> <body>" to stderr.
func PrintAPIError(step string, resp *http.Response, err error) {
	fmt.Fprintf(os.Stderr, "Error (%s): %d %s\n", step, StatusCode(resp), ErrorBody(err))
}

// Fail prints the error and exits with status 1.
func Fail(step string, resp *http.Response, err error) {
	PrintAPIError(step, resp, err)
	os.Exit(1)
}

// APIError maps an SDK failure to an HTTP status and a message for JSON responses.
// The status is taken from the API response (500 when there was none) and the
// message from the "Error" field of the body, falling back to the raw body.
func APIError(resp *http.Response, err error) (int, string) {
	status := StatusCode(resp)
	if status == 0 {
		status = http.StatusInternalServerError
	}

	body := ErrorBody(err)
	var parsed struct {
		Error string `json:"Error"`
	}
	if json.Unmarshal([]byte(body), &parsed) == nil && parsed.Error != "" {
		return status, parsed.Error
	}
	if body == "" {
		body = "Unknown error"
	}
	return status, body
}

// Elastic Email treats {...} and {{...}} in message content as template syntax,
// so strip braces from user input.
var htmlEscaper = strings.NewReplacer(
	"&", "&amp;", "<", "&lt;", ">", "&gt;", `"`, "&quot;", "'", "&#39;", "{", "&#123;", "}", "&#125;",
)

// EscapeHTML escapes a user-supplied value for an HTML body, braces included.
func EscapeHTML(s string) string {
	return htmlEscaper.Replace(s)
}

// PlainText removes braces from a user-supplied value for PlainText bodies and contact fields.
func PlainText(s string) string {
	return strings.NewReplacer("{", "", "}", "").Replace(s)
}

// HeaderText is PlainText without CR and LF, for Subject and other header-like fields.
func HeaderText(s string) string {
	return strings.NewReplacer("\r", "", "\n", "").Replace(PlainText(s))
}

// EscapeBraces turns braces into HTML entities and leaves the rest of the HTML untouched.
func EscapeBraces(s string) string {
	return strings.NewReplacer("{", "&#123;", "}", "&#125;").Replace(s)
}

// ValidAddress checks the shape of a bare address: exactly one @ and no whitespace.
func ValidAddress(email string) bool {
	return strings.Count(email, "@") == 1 && strings.IndexFunc(email, unicode.IsSpace) < 0
}

func domainOf(email string) string {
	at := strings.LastIndex(email, "@")
	if at < 0 {
		return ""
	}
	return strings.ToLower(strings.TrimSpace(strings.TrimRight(email[at+1:], "> ")))
}

// AllowedDomains is EMAIL_ALLOWED_DOMAINS, or the domain of EMAIL_TO when that is empty.
func AllowedDomains() []string {
	var domains []string
	for _, d := range strings.Split(os.Getenv("EMAIL_ALLOWED_DOMAINS"), ",") {
		if d = strings.ToLower(strings.TrimSpace(d)); d != "" {
			domains = append(domains, d)
		}
	}
	if len(domains) == 0 {
		if d := domainOf(os.Getenv("EMAIL_TO")); d != "" {
			domains = append(domains, d)
		}
	}
	return domains
}

// RecipientAllowed reports whether POST /send may deliver to this address.
func RecipientAllowed(email string) bool {
	if !ValidAddress(email) {
		return false
	}
	domain := domainOf(email)
	for _, d := range AllowedDomains() {
		if d == domain {
			return true
		}
	}
	return false
}

// BearerOk checks "Authorization: Bearer <token>" with a constant-time compare.
func BearerOk(header, token string) bool {
	const prefix = "Bearer "
	if token == "" || !strings.HasPrefix(header, prefix) {
		return false
	}
	return subtle.ConstantTimeCompare([]byte(header[len(prefix):]), []byte(token)) == 1
}

// ConfirmToken signs a double opt-in link: HMAC-SHA256(secret, email + "\n" + expires) as hex.
func ConfirmToken(secret, email string, expires int64) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(fmt.Sprintf("%s\n%d", email, expires)))
	return hex.EncodeToString(mac.Sum(nil))
}

// ConfirmURL builds a confirm link that expires after 48 hours.
func ConfirmURL(publicURL, secret, email string) string {
	expires := time.Now().Add(48 * time.Hour).Unix()
	return fmt.Sprintf("%s/double-optin/confirm?email=%s&expires=%d&token=%s",
		publicURL, url.QueryEscape(email), expires, ConfirmToken(secret, email, expires))
}

// RateLimiter counts hits per key in fixed windows.
// In-memory limits are per process; use a shared store (Redis, the platform's rate limiter) in production.
type RateLimiter struct {
	Max    int
	Window time.Duration

	mu      sync.Mutex
	windows map[string]rateWindow
}

type rateWindow struct {
	count   int
	resetAt time.Time
}

// Allow records a hit and returns false plus the seconds until the window resets when over the limit.
func (l *RateLimiter) Allow(key string) (bool, int) {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.windows == nil {
		l.windows = map[string]rateWindow{}
	}
	now := time.Now()
	w, ok := l.windows[key]
	if !ok || !now.Before(w.resetAt) {
		w = rateWindow{resetAt: now.Add(l.Window)}
	}
	w.count++
	l.windows[key] = w
	if w.count > l.Max {
		return false, int(math.Ceil(w.resetAt.Sub(now).Seconds()))
	}
	return true, 0
}

// CheckStartup logs the /send state and refuses to expose the placeholder webhook token publicly.
func CheckStartup(webhookToken string) {
	if os.Getenv("ELASTICEMAIL_SEND_TOKEN") == "" {
		log.Println("POST /send is disabled until ELASTICEMAIL_SEND_TOKEN is set")
	}
	CheckWebhookToken(webhookToken)
}

// CheckWebhookToken warns about the placeholder token locally and exits when PUBLIC_URL is public.
func CheckWebhookToken(webhookToken string) {
	if webhookToken != "" && webhookToken != "change_me" {
		return
	}
	publicURL := os.Getenv("PUBLIC_URL")
	host := ""
	if u, err := url.Parse(publicURL); err == nil {
		host = u.Hostname()
	}
	if publicURL == "" || host == "localhost" || host == "127.0.0.1" || host == "::1" {
		log.Println("ELASTICEMAIL_WEBHOOK_TOKEN is the placeholder; fine for local testing only")
		return
	}
	log.Fatalf("Refusing to start: set ELASTICEMAIL_WEBHOOK_TOKEN before exposing webhooks at %s", publicURL)
}
